pkg/scheduler/framework/runtime/batch.go

```diff
@@ -0,0 +1,236 @@
+		"profile", b.handle.ProfileName(), "cycleCount", cycleCount, "reason", reason)
+}
+
+// Update the batch state based on a the arrival of a new pod and the chosen node from the last pod.
+func (b *OpportunisticBatch) batchStateCompatible(ctx context.Context, logger klog.Logger, pod *v1.Pod, signature string, cycleCount int64, state fwk.CycleState, nodeInfos fwk.NodeInfoLister) bool {
+	// Just return if we don't have any state to use.
+	if b.stateEmpty() {
+		return false
+	}
+
+	// In this case, a previous pod was scheduled by another profile, meaning we can't use the state anymore.
+	if cycleCount != b.lastCycle.cycleCount+1 {
+		b.logUnusableState(logger, cycleCount, metrics.BatchFlushPodSkipped)
+		return false
+	}
+
+	// If our last pod failed we can't use the state.
+	if !b.lastCycle.succeeded {
+		b.logUnusableState(logger, cycleCount, metrics.BatchFlushPodFailed)
+		return false
+	}
+
+	// If the new pod is incompatible with the current state, throw the state away.
+	if signature == "" || signature != b.state.signature {
+		b.logUnusableState(logger, cycleCount, metrics.BatchFlushPodIncompatible)
+		return false
+	}
+
+	// If the state is too old, throw the state away. This is to avoid
+	// cases where we either have huge numbers of compatible pods in a
+	// row or we have a long wait between pods.
+	if time.Now().After((b.state.creationTime.Add(maxBatchAge))) {
+		b.logUnusableState(logger, cycleCount, metrics.BatchFlushExpired)
+		return false
+	}
+
+	// We can only reuse the previous state if the new pod will not
+	// fit on the node we used for the last pod. If the node we
+	// chose last time can't host this pod, then we know
+	// that the next best should be the next node in the list.
+	// If we *could* put this pod on the node we just used, then we can't
+	// use our cache because we don't know how to rescore the used node; it
+	// could be the best, or it could be somewhere else.
+	// See git.k8s.io/enhancements/keps/sig-scheduling/5598-opportunistic-batching
+	lastChosenNode, err := nodeInfos.Get(b.lastCycle.chosenNode)
+	if lastChosenNode == nil || err != nil {
+		b.logUnusableState(logger, cycleCount, metrics.BatchFlushNodeMissing)
+		return false
+	}
+
+	status := b.handle.RunFilterPlugins(ctx, state, pod, lastChosenNode)
+	if !status.IsRejected() {
+		b.logUnusableState(logger, cycleCount, metrics.BatchFlushNodeNotFull)
+		return false
+	}
+
+	// Our state matches with our new pod and is useable
+	return true
+}
+
+// Irritatingly we can end up with a variety of different configurations that are all "empty".
+// Rather than trying to normalize all cases when they happen, we just check them all.
+func (b *OpportunisticBatch) stateEmpty() bool {
+	return b.state == nil || b.state.sortedNodes == nil || b.state.sortedNodes.Len() == 0
+}
+
+func newOpportunisticBatch(h fwk.Handle, signatureFunc PodSignatureFunc) *OpportunisticBatch {
+	return &OpportunisticBatch{
+		signatureFunc: signatureFunc,
```

Comment:
You added SignPod onto fwk.Handle. Why do we need to pass signatureFunc? Can we just run SignPod through fwk.Handle?